petab/simplify.py

```diff
@@ -0,0 +1,106 @@
+"""Functionality for simplifying PEtab problems"""
+from math import nan
+
+import pandas as pd
+
+from . import Problem
+from .C import *  # noqa: F403
+from .lint import lint_problem
+
+__all__ = [
+    "remove_nan_measurements",
+    "remove_unused_observables",
+    "remove_unused_conditions",
+    "simplify_problem",
+    "condition_parameters_to_parameter_table",
+]
+
+
+def remove_nan_measurements(problem: Problem):
+    """Drop any measurements that are NaN"""
+    problem.measurement_df = problem.measurement_df[
+        ~problem.measurement_df[MEASUREMENT].isna()
+    ]
+    problem.measurement_df.reset_index(inplace=True, drop=True)
+
+
+def remove_unused_observables(problem: Problem):
+    """Remove observables that have no measurements"""
+    measured_observables = set(problem.measurement_df[OBSERVABLE_ID].unique())
+    problem.observable_df = problem.observable_df[
+        problem.observable_df.index.isin(measured_observables)]
+
+
+def remove_unused_conditions(problem: Problem):
+    """Remove conditions that have no measurements"""
+    measured_conditions = \
+        set(problem.measurement_df[SIMULATION_CONDITION_ID].unique())
+    if PREEQUILIBRATION_CONDITION_ID in problem.measurement_df:
+        measured_conditions |= \
+            set(problem.measurement_df[PREEQUILIBRATION_CONDITION_ID].unique())
+
+    problem.condition_df = problem.condition_df[
+        problem.condition_df.index.isin(measured_conditions)]
+
+
+def simplify_problem(problem: Problem):
+    if lint_problem(problem):
+        raise ValueError("Invalid PEtab problem supplied.")
+
+    remove_unused_observables(problem)
+    remove_unused_conditions(problem)
+    condition_parameters_to_parameter_table(problem)
+
+    if lint_problem(problem):
+        raise AssertionError("Invalid PEtab problem generated.")
+
+
+def condition_parameters_to_parameter_table(problem: Problem):
+    """Move parameters from the condition table to the parameters table, if
+    the same parameter value is used for all conditions."""
+    if problem.condition_df is None or problem.condition_df.empty \
+            or problem.model is None:
+        return
+
+    replacements = {}
+    for parameter_id in problem.condition_df:
+        if parameter_id == PARAMETER_NAME:
+            continue
+
+        if problem.model.is_state_variable(parameter_id):
+            # initial states can't go the parameters table
+            continue
+
+        series = problem.condition_df[parameter_id]
+        if series.dtype == 'o':
+            # We don't touch parametric overrides
+            continue
+
+        # same value for all conditions?
+        if len(series.unique()) == 1:
+            replacements[parameter_id] = series[0]
```

Comment:
Is it always `'o'` for incompatible columns, e.g. not `str`? Could a column of floats be `'o'` anyway? A possible alternative.
```suggestion
        values = list(set(problem.condition_df[parameter_id]))
        if len(values) != 1:
            continue

        try:
            float_value = float(values[0])
            assert float_value == values[0]
        except:
            continue

        replacements[parameter_id] = float_value
```

Reply:
Yes, `str` would also be `O`. But you could convert an all-floats series to `O` too, right.

Reply:
Done